torch/distributions/multivariate_normal.py

```diff
@@ -0,0 +1,154 @@
+import math
+from numbers import Number
+
+import torch
+from torch.autograd import Variable
+from torch.distributions import constraints
+from torch.distributions.distribution import Distribution
+from torch.distributions.utils import broadcast_all, lazy_property
+
+
+def _get_batch_shape(bmat, bvec):
+    """
+    Given a batch of matrices and a batch of vectors, compute the combined `batch_shape`.
+    """
+    try:
+        vec_shape = torch._C._infer_size(bvec.shape, bmat.shape[:-1])
+    except RuntimeError:
+        raise ValueError("Incompatible batch shapes: vector {}, matrix {}".format(bvec.shape, bmat.shape))
+    return torch.Size(vec_shape[:-1])
+
+
+def _batch_mv(bmat, bvec):
+    """
+    Performs a batched matrix-vector product, with an arbitrary batch shape.
+    """
+    batch_shape = bvec.shape[:-1]
+    event_dim = bvec.shape[-1]
+    bmat = bmat.expand(batch_shape + (event_dim, event_dim))
+    if batch_shape != bmat.shape[:-2]:
+        raise ValueError("Batch shapes do not match: matrix {}, vector {}".format(bmat.shape, bvec.shape))
+    bvec = bvec.unsqueeze(-1)
```

Comment:
Why do we need all this broadcasting here? Doesn't `torch.bmm` do it?

Reply:
To call `torch.bmm`, as I understand it we need both arguments to be 3D tensors. If the batch_shape is multidimensional, then this flattens it down. It's more complicated here than elsewhere because the matrix `bmat` might be missing leading dimensions of the batch shape.

I wanted to avoid pre-emptively calling `.expand` on the `scale_tril` matrix, since in the `_batch_mahalanobis` helper (line 71) we split along the batch dimension, call `torch.inverse`, and then call `torch.stack`. 

This whole problem comes about because the leading dimensions of the`loc` and `scale_tril` parameters might not be the same. If the batch size is large (e.g. because loc has a batch dimension) but there is only a single covariance matrix, then we don't mistakenly end up inverting the single covariance matrix once for every item in the batch. (Originally we had also discussed prohibiting any batching of the `covariance_matrix` or `scale_tril` parameter.)

If I understand correctly, I think we could probably remove this entire workaround after #4612.

Reply:
Can you please document what are the expected sizes of inputs to this function? It's very hard to review it without checking and fully understanding all the places where it's used.

Reply:
Apologies — this was really far more complicated than it needed to be. I rewrote (and documented) this helper, and it should be much more clear now.